scripts/xsslint/xsslint/reporting.py

```diff
@@ -0,0 +1,376 @@
+        Returns a key that can be sorted on
+        """
+        return (0, 0, self.rule.rule_id)
+
+    def first_line(self):
+        """
+        Since a file level rule has no first line, returns empty string.
+        """
+        return ''
+
+    def prepare_results(self, full_path, string_lines):
+        """
+        Preps this instance for results reporting.
+
+        Arguments:
+            full_path: Path of the file in violation.
+            string_lines: A StringLines containing the contents of the file in
+                violation.
+
+        """
+        self.full_path = full_path
+        self._mark_disabled(string_lines.get_string())
+
+    def print_results(self, _options, out):
+        """
+        Prints the results represented by this rule violation.
+
+        Arguments:
+            _options: ignored
+            out: output file
+        """
+        print("{}: {}".format(self.full_path, self.rule.rule_id), file=out)
+
+
+class ExpressionRuleViolation(RuleViolation):
+    """
+    A class representing a particular rule violation for expressions which
+    contain more specific details of the location of the violation for reporting
+    purposes.
+
+    """
+
+    def __init__(self, rule, expression):
+        """
+        Init method.
+
+        Arguments:
+            rule: The Rule which was violated.
+            expression: The Expression that was in violation.
+
+        """
+        super(ExpressionRuleViolation, self).__init__(rule)
+        self.expression = expression
+        self.start_line = 0
+        self.start_column = 0
+        self.end_line = 0
+        self.end_column = 0
+        self.lines = []
+        self.is_disabled = False
+
+    def _mark_expression_disabled(self, string_lines):
+        """
+        Marks the expression violation as disabled if it finds the disable
+        pragma anywhere on the first line of the violation, or at the start of
+        the line preceding the violation.
+
+        Pragma format::
+
+            xss-lint: disable=violation-name,other-violation-name
+
+        Examples::
+
+            <% // xss-lint: disable=underscore-not-escaped %>
+            <%= gettext('Single Line') %>
+
+            <%= gettext('Single Line') %><% // xss-lint: disable=underscore-not-escaped %>
+
+        Arguments:
+            string_lines: A StringLines containing the contents of the file in
+                violation.
+
+        Side Effect:
+            Sets self.is_disabled as appropriate based on whether the pragma is
+            found.
+
+        """
+        # disable pragma can be at the start of the preceding line
+        has_previous_line = self.start_line > 1
+        if has_previous_line:
+            line_to_check = string_lines.line_number_to_line(self.start_line - 1)
+            self._mark_disabled(line_to_check, scope_start_string=True)
+            if self.is_disabled:
+                return
+
+        # TODO: this should work at end of any line of the violation
```

Comment:
May be a separate follow-up PR/ticket, but we should audit TODO comments and remove or add tickets where necessary.  This is one that should just be dropped.  I wouldn't do it as part of this PR, because I think we should keep this refactor as clean as possible.

Reply:
Ok